pkg/kinflate/transformers/overlay.go

```diff
@@ -59,39 +65,143 @@ func (o *overlayTransformer) Transform(baseResourceMap resource.ResourceCollecti
 		switch {
 		case runtime.IsNotRegisteredError(err):
 			// Use JSON merge patch to handle types w/o schema
-			baseBytes, err := json.Marshal(base.Data)
+			merged, err = jsonMergePatch(base.Data.Object, overlay.Data.Object, jsonpatch.MergePatch)
 			if err != nil {
 				return err
 			}
-			patchBytes, err := json.Marshal(overlay.Data)
+		case err != nil:
+			return err
+		default:
+			// Use Strategic Merge Patch to handle types w/ schema
+			// TODO: Change this to use the new Merge package.
+			// Store the name of the base object, because this name may have been munged.
+			// Apply this name to the StrategicMergePatched object.
+			lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(versionedObj)
 			if err != nil {
 				return err
 			}
-			mergedBytes, err := jsonpatch.MergePatch(baseBytes, patchBytes)
+			merged, err = strategicpatch.StrategicMergeMapPatchUsingLookupPatchMeta(
+				base.Data.Object,
+				overlay.Data.Object,
+				lookupPatchMeta)
 			if err != nil {
 				return err
 			}
-			err = json.Unmarshal(mergedBytes, &merged)
+		}
+		base.Data.SetName(baseName)
+		baseResourceMap[gvkn].Data.Object = merged
+	}
+	return nil
+}
+
+// mergePatches merge and index patches by GVKN.
+// It errors out if there is conflict between patches.
+func (o *overlayTransformer) mergePatches() (resource.ResourceCollection, error) {
+	rc := resource.ResourceCollection{}
+	for _, patch := range o.overlay {
+		gvkn := patch.GVKN()
+		existing, found := rc[gvkn]
+		if !found {
+			rc[gvkn] = patch
+			continue
+		}
+		var merged map[string]interface{}
+		versionedObj, err := scheme.Scheme.New(gvkn.GVK)
+		switch {
+		case runtime.IsNotRegisteredError(err):
```

Comment:
On a completely unrelated note, why do we need to have the types registered to use smp?

Reply:
SMP need the scheme to generate a versioned object for reference.
We can easily switch it to use OpenAPI. I have the code ready, but we need to convince people we need to contact the server.